src/contracts/core/AVSDirectory.sol

```diff
@@ -419,6 +521,196 @@ contract AVSDirectory is
         }
     }
 
+    /**
+     * @notice For a single strategy, update freeMagnitude by adding completable pending free magnitudes
+     * @param operator address to update freeMagnitude for
+     * @param strategy the strategy to update freeMagnitude for
+     * @param numToComplete the number of pending free magnitudes deallocations to complete, 0 to complete all (uint8 max 256)
+     * @dev read through pending free magnitudes and add to freeMagnitude if completableTimestamp is >= block timestamp
+     * In addition to updating freeMagnitude, updates next starting index to read from for pending free magnitudes after completing
+     */
+    function _updateFreeMagnitude(address operator, IStrategy strategy, uint8 numToComplete) internal {
+        (uint64 freeMagnitudeToAdd, uint256 nextIndex) = _getPendingFreeMagnitude(operator, strategy, numToComplete);
+        freeMagnitude[operator][strategy] += freeMagnitudeToAdd;
+        _nextPendingFreeMagnitudeIndex[operator][strategy] = nextIndex;
+    }
+
+    /**
+     * @notice For a single strategy, modify magnitude allocations for each of the specified operatorSets
+     * @param operator address to modify allocations for
+     * @param allocation the magnitude allocations to modify for a single strategy
+     * @param allocationEffectTimestamp the timestamp when the allocations will take effect
+     * @param deallocationCompletableTimestamp the timestamp when the deallocations will be completable
+     */
+    function _modifyAllocations(
+        address operator,
+        MagnitudeAllocation calldata allocation,
+        uint32 allocationEffectTimestamp,
+        uint32 deallocationCompletableTimestamp
+    ) internal {
+        uint64 newFreeMagnitude = freeMagnitude[operator][allocation.strategy];
+        OperatorSet[] calldata opSets = allocation.operatorSets;
+
+        for (uint256 i = 0; i < opSets.length; ++i) {
+            /// TODO??? ensure ordered in ascending bytes32 hash of opSets, can we have duplicate opSets, duplicate strategies?
```

Comment:
check that operator set exists

Reply:
https://github.com/Layr-Labs/eigenlayer-contracts/pull/679/files#diff-0b976d6e833faec8371ead1185950cd0af501fe77ba645a0478e70d7e220c85cR266-R269